optimum/habana/transformers/models/esm/modeling_esmfold.py

```diff
@@ -0,0 +1,438 @@
+# coding=utf-8
+# Copyright 2022 Meta and The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Optional
+
+import numpy as np
+import torch
+
+# from .configuration_esm import EsmConfig
+from transformers.models.esm.modeling_esmfold import EsmForProteinFoldingOutput, categorical_lddt, flatten_final_dims
+from transformers.models.esm.openfold_utils import (
+    compute_predicted_aligned_error,
+    compute_tm,
+    make_atom14_masks,
+)
+
+# from ...modeling_outputs import ModelOutput
+from transformers.utils import (
+    ContextManagers,
+)
+
+
+def gaudi_esmfolding_trunk_forward(self, seq_feats, pair_feats, true_aa, residx, mask, no_recycles):
+    """
+    Inputs:
+        seq_feats: B x L x C tensor of sequence features pair_feats: B x L x L x C tensor of pair features residx: B
+        x L long tensor giving the position in the sequence mask: B x L boolean tensor indicating valid residues
+
+    Output:
+        predicted_structure: B x L x (num_atoms_per_residue * 3) tensor wrapped in a Coordinates object
+
+    Copied from EsmFoldingTrunk.forward:
+    https://github.com/huggingface/transformers/blob/main/src/transformers/models/esm/modeling_esmfold.py
+    The change is:
+    - Add extra mark_step in trunk_iter for each block.
+    """
+
+    device = seq_feats.device
+    s_s_0 = seq_feats
+    s_z_0 = pair_feats
+
+    if no_recycles is None:
+        no_recycles = self.config.max_recycles
+    else:
+        if no_recycles < 0:
+            raise ValueError("Number of recycles must not be negative.")
+        no_recycles += 1  # First 'recycle' is just the standard forward pass through the model.
+
+    def trunk_iter(s, z, residx, mask):
+        z = z + self.pairwise_positional_embedding(residx, mask=mask)
+
+        for block in self.blocks:
+            s, z = block(s, z, mask=mask, residue_index=residx, chunk_size=self.chunk_size)
+            if s.device.type == "hpu":
```

Comment:
I prefer to skip the check of lazy mode here because
1. As you said, eager mode is more for debugging and lazy mode is the default mode when user running it.
2. Even under the eager mode, mark_step should work. It will be basically a NOP under eager mode.
3. Adding lazy mode check here makes code clumsy :)

Reply:
Sounds good :+1: